README.md
node-dig
========

node-dig is a library based on **tjfontaine/native-dns-packet** that enable the more powerful ability to manipulate DNS request, comparing to the original `require("dns")` module, it exposed the DNS underlying mechanism to Node.js layer. It's easy to build customized DNS utilities base on it, such as Unix command `dig`.

### What's different?

We modified the packet generator, made it possible to customize DNS Header to be sent. (for the purpose that to support `[no]Recursion Desired` in previous, but original node-dns has hard-coded the `RD` bit to `1`, so we hack it).

```coffee
# supported DNS request header fields

# jump to https://www.ietf.org/rfc/rfc1035.txt (4.1.1. Header section format) for a detail view.
qr
opcode
aa
tc
rd
ra
res1
res2
res3
rcode
```

Installation
------------

```
npm i node-dig
```

Test

```
npm test
```

Usages
-------

```js
var dig = require('node-dig');

var question = dig.Question({
  name: 'www.google.com.co',
  type: 'NS',
});

var req = dig
  .Request({
    question: question,
    timeout: 1000,
    // customize header
    header: {
      // no Recursion Desired
      rd: 0
    }
  })
  .on('timeout', function () {
    console.log('Timeout in making request');
  })
  .on('message', function (err, answer) {
    answer.answer.forEach(function (a) {
      console.log(a.address);
    });
  })
  .on('end', function () {
    console.log('end query');
  })

req.send();
```

API
---

## Class: `Question`

```js
var question = dig.Question(queryType);
```

## Class: `Request`

```js
var request = require('node-dig').Request(options);
```

There are only two methods on `request`

 * `request.send` -- sends the actual request to the remote endpoint
 * `request.cancel` -- cancels the request and ignores any responses

QueryType
---------

queryTypes is an object with the following fields:

+ `name`: server name
+ `type`: query type

the `type` list below are available:

```
'A'
'AAAA'
'NS'
'CNAME'
'PTR'
'NAPTR'
'TXT'
'MX'
'SRV'
'SOA'
'TLSA'
```

Options
-------

Request creation takes an object with the following fields

 * `question` -- an instance of Question (required)
 * `server` -- defines the remote end point (optional)
  - as an object it should be
    * `address` -- a string ip address (required)
    * `port` -- a number for the remote port (optional, default 53)
    * `type` -- a string indicating `udp` or `tcp` (optional, default `udp`)
You do not need to indicate ipv4 or ipv6, the backend will handle that,
if you hadn't specified this field, it'll be setted to system wide `nameserver` from `/etc/resolv.conf`
  - a string ip address
 * `header` -- DNS Header set(optional), see rfc1035 for details.
 * `timeout` -- a number in milliseconds indicating how long to wait for the
request to finish. (optional, default 4000)

Events type
-----------

Request emits the following events

 * `message` -- This is where you get a response, passes `(err, answer)` where
answer is an instance of `Packet`
 * `timeout` -- Fired when the timeout is reached
 * `cancelled` -- Fired if the request is cancelled
 * `end` -- Always fired after a request finished, regardless of disposition

Mock
----

we support mock by mounting an mock settings object on the exposed module.

### Usage

```js
var dig = require('node-dig');

// into mock env
dig.mock = {
  timeout: true
};

// main logic goes here...

// jump out of mock env any time
dig.mock = null;

// continue ...
```

### mock settings object

The following fields are available:

+ `timeout`: <Boolean>
+ `end`: <Boolean>
+ `message`: <Object>
+ `cancelled`: <Boolean>


History
0.0.3 / 2016-08-16
==================

+ 规范化目录结构
+ 删除无用示例和测试

0.0.2 / 2015-12-22
==================

+ lib/client.js: clean up
+ lib/client.js: only emit "timeout" when not setting or setted to be true
+ lib/client.js: fix a logical bug
+ index.js: 使用setter定义mock attribute
+ README: 增加mock描述
+ index.js | lib/client.js: 添加mock环境选项
+ README.md: 添加request header field支持说明
+ lib/pending.js: 去除try_edns字段
+ lib/client.js: 去掉封装函数, 仅保留了核心的Request对象, 去掉不必
要的cache功能
+ lib/utils.js: 添加`/etc/resolv.conf`解析器
+ dns.js: 取消封装功能的支持
+ lib/*: 支持自定义DNS报文Header部分的值

test/packet.js
var dns = require('..'),
  Packet = require('../lib/packet');

exports.roundTrip = function(test) {
  var buff, pre, post, len;

  pre = new Packet();
  pre.header.id = 12345;
  pre.header.rcode = 1;

  pre.question.push(dns.Question({
    name: 'www.google.com',
    type: dns.consts.NAME_TO_QTYPE.A
  }));

  pre.answer.push(dns.A({
    name: 'www.google.com',
    address: '127.0.0.1',
    ttl: 600
  }));

  buff = new Buffer(1024);

  len = Packet.write(buff, pre);

  post = Packet.parse(buff.slice(0, len));
  delete pre.address;

  test.deepEqual(pre, post);
  test.done();
};

// exports.truncate = function(test) {
//   var buff, pre, post, i;

//   pre = new Packet();
//   pre.header.id = 12345;
//   pre.header.rcode = 1;

//   pre.question.push(dns.Question({
//     name: 'really.long.name.some.domain.com',
//     type: 'A',
//   }));

//   for (i = 0; i < 254; i ++) {
//     pre.answer.push(dns.A({
//       name: i+'.'+i+'.'+i+'.really.long.name.some.domain.com',
//       address: '127.0.0.' + i,
//       ttl: 600,
//     }));
//     pre.authority.push(dns.A({
//       name: i+'.'+i+'.'+i+'.really.long.name.some.domain.com',
//       address: '127.0.0.' + i,
//       ttl: 600,
//     }));
//     pre.additional.push(dns.A({
//       name: i+'.'+i+'.'+i+'.really.long.name.some.domain.com',
//       address: '127.0.0.' + i,
//       ttl: 600,
//     }));
//   }

//   buff = new Buffer(512);
//   len = Packet.write(buff, pre);
//   post = Packet.parse(buff.slice(0, len));

//   test.notEqual(pre.additional.length, post.additional.length,
//     'Additional should be less because of truncated packet');

//   test.done();
// };


test/request.js
var dns = require('..'),
  Request = dns.Request,
  Question = dns.Question,
  consts = dns.consts,
  dgram = require('dgram');

var q = Question({
  name: 'www.google.com',
  type: consts.NAME_TO_QTYPE.A
});

var noServer = {
  address: '127.0.0.1',
  port: 19999,
  type: 'udp'
};

var udpServ = {
  address: '8.8.8.8',
  port: 53,
  type: 'udp'
};

var tcpServ = {
  address: '8.8.8.8',
  port: 53,
  type: 'tcp'
};

exports.setUp = function(cb) {
  cb();
};

exports.timeout = function(test) {
  var r = Request({
    question: q,
    server: noServer,
    timeout: 100,
    cache: false
  });

  var timedout = false;

  r.on('timeout', function() {
    timedout = true;
  });

  r.on('end', function() {
    test.ok(timedout, 'Failed to timeout');
    test.done();
  });

  r.send();
};

exports.udpResponse = function(test) {
  var r = Request({
    question: q,
    server: udpServ,
    timeout: 4000
  });

  r.on('message', function(err, answer) {
    test.ok(!err, 'UDP Request should not error');
    test.ok(answer, 'No UDP answer provided');
    test.ok(answer.answer.length > 0, 'No answers found');
  });

  r.on('timeout', function() {
    test.ok(false, 'UDP Requests should not timeout');
  });

  r.on('end', function() {
    test.done();
  });

  r.send();
};

exports.tcpResponse = function(test) {
  var r = Request({
    question: q,
    server: tcpServ,
    timeout: 4000
  });

  r.on('message', function(err, answer) {
    test.ok(!err, 'TCP Request should not error');
    test.ok(answer, 'No TCP answer provided');
    test.ok(answer.answer.length > 0, 'No answers found');
  });

  r.on('timeout', function() {
    test.ok(false, 'TCP Requests should not timeout');
  });

  r.on('end', function() {
    test.done();
  });

  r.send();
};

exports.serverString = function(test) {
  var r = Request({
    question: q,
    server: '8.8.8.8'
  });

  r.on('message', function(err, answer) {
    test.ok(answer.answer.length > 0, 'no answers found');
  });

  r.on('timeout', function() {
    test.ok(false, 'Should not timeout');
  });

  r.on('end', function() {
    test.done();
  });

  r.send();
};

exports.questionString = function(test) {
  var r = Request({
    question: Question({
      name: 'www.google.com',
      type: 'a'
    }),
    server: '8.8.8.8'
  });

  r.on('message', function(err, answer) {
    test.ok(answer.answer.length > 0, 'no answers found');
  });

  r.on('timeout', function() {
    test.ok(false, 'Should not timeout');
  });

  r.on('end', function() {
    test.done();
  });

  r.send();
};

exports.emptyUdp = function(test) {
  var socket = dgram.createSocket('udp4');
  socket.on('listening', function() {
    var timeout = false;
    var r = Request({
      question: q,
      server: { address: '127.0.0.1', port: socket.address().port, type: 'udp' },
      timeout: 300,
      cache: false
    });
    r.on('message', function() {
      test.ok(false, 'There should not be a response');
    });
    r.on('timeout', function() {
      timeout = true;
    });
    r.on('end', function() {
      test.ok(timeout, 'This should timeout');
      socket.close();
      test.done();
    });
    r.send();
  });
  socket.on('message', function(msg, remote) {
    socket.send(new Buffer(1), 0, 1, remote.port, remote.address);
  });
  socket.bind();
};

exports.longName = function(test) {
  var didErr = false;
  var r = Request({
    question: Question({
      name: '*************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************' +
        '***************'
    }),
    server: '8.8.8.8',
    timeout: 1000
  });
  r.on('error', function(err) {
    test.ok(err, 'We should error because the packet failed to pack');
    didErr = true;
  });
  r.on('end', function() {
    test.ok(!didErr, 'We did not err');
    test.done();
  });
  r.send();
};

exports.tearDown = function(cb) {
  cb();
};


dns.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

"use strict";

exports.platform = require('./lib/platform');

var client = require('./lib/client');
exports.lookup = client.lookup;
exports.resolve = client.resolve;
exports.resolve4 = client.resolve4;
exports.resolve6 = client.resolve6;
exports.resolveMx = client.resolveMx;
exports.resolveTxt = client.resolveTxt;
exports.resolveSrv = client.resolveSrv;
exports.resolveNs = client.resolveNs;
exports.resolveCname = client.resolveCname;
exports.reverse = client.reverse;

var consts = require('native-dns-packet').consts;
exports.BADNAME = consts.BADNAME;
exports.BADRESP = consts.BADRESP;
exports.CONNREFUSED = consts.CONNREFUSED;
exports.DESTRUCTION = consts.DESTRUCTION;
exports.REFUSED = consts.REFUSED;
exports.FORMERR = consts.FORMERR;
exports.NODATA = consts.NODATA;
exports.NOMEM = consts.NOMEM;
exports.NOTFOUND = consts.NOTFOUND;
exports.NOTIMP = consts.NOTIMP;
exports.SERVFAIL = consts.SERVFAIL;
exports.TIMEOUT = consts.TIMEOUT;
exports.consts = consts;

var definedTypes = [
  'A',
  'AAAA',
  'NS',
  'CNAME',
  'PTR',
  'NAPTR',
  'TXT',
  'MX',
  'SRV',
  'SOA',
  'TLSA',
].forEach(function (type) {
  exports[type] = function (opts) {
    var obj = {};
    opts = opts || {};
    obj.type = consts.nameToQtype(type);
    obj.class = consts.NAME_TO_QCLASS.IN;
    Object.keys(opts).forEach(function (k) {
      if (opts.hasOwnProperty(k) && ['type', 'class'].indexOf(k) == -1) {
        obj[k] = opts[k];
      }
    });
    return obj;
  };
});

exports.Question = function (opts) {
  var q = {}, qtype;

  opts = opts || {};

  q.name = opts.name;

  qtype = opts.type || consts.NAME_TO_QTYPE.A;
  if (typeof(qtype) === 'string' || qtype instanceof String)
    qtype = consts.nameToQtype(qtype.toUpperCase());

  if (!qtype || typeof(qtype) !== 'number')
    throw new Error("Question type must be defined and be valid");

  q.type = qtype;

  q.class = opts.class || consts.NAME_TO_QCLASS.IN;

  return q;
};
exports.Request = client.Request;


index.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

'use strict';

var client = require('./lib/client');
// exports.lookup = client.lookup;
// exports.resolve = client.resolve;
// exports.resolve4 = client.resolve4;
// exports.resolve6 = client.resolve6;
// exports.resolveMx = client.resolveMx;
// exports.resolveTxt = client.resolveTxt;
// exports.resolveSrv = client.resolveSrv;
// exports.resolveNs = client.resolveNs;
// exports.resolveCname = client.resolveCname;
// exports.reverse = client.reverse;

var consts = require('native-dns-packet').consts;
exports.BADNAME = consts.BADNAME;
exports.BADRESP = consts.BADRESP;
exports.CONNREFUSED = consts.CONNREFUSED;
exports.DESTRUCTION = consts.DESTRUCTION;
exports.REFUSED = consts.REFUSED;
exports.FORMERR = consts.FORMERR;
exports.NODATA = consts.NODATA;
exports.NOMEM = consts.NOMEM;
exports.NOTFOUND = consts.NOTFOUND;
exports.NOTIMP = consts.NOTIMP;
exports.SERVFAIL = consts.SERVFAIL;
exports.TIMEOUT = consts.TIMEOUT;
exports.consts = consts;

[
  'A',
  'AAAA',
  'NS',
  'CNAME',
  'PTR',
  'NAPTR',
  'TXT',
  'MX',
  'SRV',
  'SOA',
  'TLSA'
].forEach(function(type) {
  exports[type] = function(opts) {
    var obj = {};
    opts = opts || {};
    obj.type = consts.nameToQtype(type);
    obj.class = consts.NAME_TO_QCLASS.IN;
    Object.keys(opts).forEach(function(k) {
      if (opts.hasOwnProperty(k) && ['type', 'class'].indexOf(k) == -1) {
        obj[k] = opts[k];
      }
    });
    return obj;
  };
});

exports.Question = function(opts) {
  var q = {}, qtype;

  opts = opts || {};

  q.name = opts.name;

  qtype = opts.type || consts.NAME_TO_QTYPE.A;
  if (typeof qtype === 'string' || qtype instanceof String)
    qtype = consts.nameToQtype(qtype.toUpperCase());

  if (!qtype || typeof qtype !== 'number')
    throw new Error('Question type must be defined and be valid');

  q.type = qtype;

  q.class = opts.class || consts.NAME_TO_QCLASS.IN;

  return q;
};
exports.Request = client.Request;

// support mock test
Object.defineProperty(exports, 'mock', {
  set: function(mockOptions) {
    this.Request.mock = false;
    if (mockOptions)
      this.Request.mock = mockOptions;
  }
});


lib/client.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

'use strict';

var net = require('net');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var PendingRequests = require('./pending');
var utils = require('./utils');

var debug = function() {};

if (process.env.NODE_DEBUG && process.env.NODE_DEBUG.match(/dns/)) {
  debug = function debug() {
    var args = Array.prototype.slice.call(arguments);
    console.error.apply(this, ['client', Date.now().toString()].concat(args));
  };
}

var Request = exports.Request = function(opts) {
  if (!(this instanceof Request)) return new Request(opts);

  this.question = opts.question;
  this.server = opts.server;

  // additional settings: header
  this.header = opts.header || {};

  if (typeof this.server === 'string' || this.server instanceof String)
    this.server = { address: this.server, port: 53, type: 'udp' };

  if (!this.server || !this.server.address || !net.isIP(this.server.address))
    // use platform specified
    this.server = { address: utils.parseResolv()[0] };

  if (!this.server.type || ['udp', 'tcp'].indexOf(this.server.type) === -1)
    this.server.type = 'udp';

  if (!this.server.port)
    this.server.port = 53;

  this.timeout = opts.timeout || 4 * 1000;
  this.fired = false;
  this.id = undefined;

  debug('request created', this.question);
};

util.inherits(Request, EventEmitter);

// mock settings
Request.mock = false;

Request.prototype.handle = function(err, answer, cached) {
  var mock = Request.mock;

  if (!this.fired) {
    debug('request handled', this.id, this.question);

    if (mock && mock.message === false) {
      return this.done();
    }

    if (mock && mock.message) {
      answer = mock.message;
    }

    this.emit('message', err, answer);
    this.done();
  }
};

Request.prototype.done = function() {
  var mock = Request.mock;

  debug('request finished', this.id, this.question);
  this.fired = true;
  clearTimeout(this.timer_);
  PendingRequests.remove(this);

  if (!mock || !('end' in mock) || mock.end === true) {
    this.emit('end');
  }

  this.id = undefined;
};

Request.prototype.handleTimeout = function() {
  var mock = Request.mock;

  if (!this.fired) {
    debug('request timedout', this.id, this.question);

    if (!mock || !('timeout' in mock) || mock.timeout === true)
      this.emit('timeout');

    this.done();
  }
};

Request.prototype.error = function(err) {
  if (!this.fired) {
    debug('request error', err, this.id, this.question);
    this.emit('error', err);
    this.done();
  }
};

Request.prototype.send = function() {
  var mock = Request.mock;

  debug('request not in cache', this.question);
  var self = this;

  if (mock && mock.timeout === true) {
    self.handleTimeout();
  } else {
    self.timer_ = setTimeout(function() {
      self.handleTimeout();
    }, self.timeout);

    PendingRequests.send(self);
  }
};

Request.prototype.cancel = function() {
  var mock = Request.mock;

  debug('request cancelled', this.id, this.question);

  if (!mock || !('cancelled' in mock) || mock.cancelled === true) {
    this.emit('cancelled');
    this.done();
  }
};


lib/packet.js
// Copyright 2011 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the 'Software'), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

'use strict';

var NDP = require('native-dns-packet');
var util = require('util');

var Packet = module.exports = function(socket) {
  NDP.call(this);
  this.address = undefined;
  this._socket = socket;
};
util.inherits(Packet, NDP);

Packet.prototype.send = function() {
  var buff, len, size;

  if (typeof this.edns_version !== 'undefined') {
    size = 4096;
  }

  this.payload = size = size || this._socket.base_size;

  buff = this._socket.buffer(size);
  len = Packet.write(buff, this);
  this._socket.send(len);
};

Packet.parse = function(msg, socket) {
  var p = NDP.parse(msg);
  p._socket = socket;
  return p;
};

Packet.write = NDP.write;


lib/pending.js
// Copyright 2012 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

'use strict';

var net = require('net');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var Packet = require('./packet');
var consts = require('native-dns-packet').consts;
var UDPSocket = require('./utils').UDPSocket;
var TCPSocket = require('./utils').TCPSocket;

var debug = function() {};

var SocketQueue = function(socket, server) {
  this._active = {};
  this._active_count = 0;
  this._pending = [];

  debug('created', server);

  this._server = server;

  this._socket = socket;
  this._socket.on('ready', this._onlisten.bind(this));
  this._socket.on('message', this._onmessage.bind(this));
  this._socket.on('close', this._onclose.bind(this));
  this._socket.bind(server);

  this._refd = true;
};
util.inherits(SocketQueue, EventEmitter);

SocketQueue.prototype.send = function(request) {
  debug('added', request.question);
  this._pending.push(request);
  this._fill();
};

SocketQueue.prototype.remove = function(request) {
  var req = this._active[request.id];
  var idx = this._pending.indexOf(request);

  if (req) {
    delete this._active[request.id];
    this._active_count -= 1;
    this._fill();
  }

  if (idx > -1)
    this._pending.splice(idx, 1);

  this._unref();
};

SocketQueue.prototype.close = function() {
  debug('closing', this._server);
  this._socket.close();
  this._socket = undefined;
  this.emit('close');
};

SocketQueue.prototype._fill = function() {
  debug('pre fill, active:', this._active_count, 'pending:',
        this._pending.length);

  while (this._listening && this._pending.length &&
         this._active_count < 100) {
    this._dequeue();
  }

  debug('post fill, active:', this._active_count, 'pending:',
        this._pending.length);
};

var randomInteger = function() {
  return Math.floor(Math.random() * 50000 + 1);
};

SocketQueue.prototype._dequeue = function() {
  var req = this._pending.pop();
  var id, packet;

  if (req) {
    id = randomInteger();

    while (this._active[id])
      id = randomInteger();

    debug('sending', req.question, id);

    req.id = id;
    this._active[id] = req;
    this._active_count += 1;

    try {
      packet = new Packet(this._socket.remote(req.server));

      packet.header.id = id;
      Object.assign(packet.header, req.header);

      packet.question.push(req.question);
      packet.send();

      this._ref();
    } catch (e) {
      req.error(e);
    }
  }
};

SocketQueue.prototype._onmessage = function(msg, remote) {
  var req, packet;

  debug('got a message', this._server);

  try {
    packet = Packet.parse(msg, remote);
    req = this._active[packet.header.id];
    debug('associated message', packet.header.id);
  } catch (e) {
    debug('error parsing packet', e);
  }

  if (req) {
    delete this._active[packet.header.id];
    this._active_count -= 1;
    req.handle(null, packet);
    this._fill();
  }

  this._unref();
};

SocketQueue.prototype._unref = function() {
  var self = this;
  this._refd = false;

  if (this._active_count <= 0) {
    if (this._socket.unref) {
      debug('unrefd socket');
      this._socket.unref();
    } else if (!this._timer) {
      this._timer = setTimeout(function() {
        self.close();
      }, 300);
    }
  }
};

SocketQueue.prototype._ref = function() {
  this._refd = true;
  if (this._socket.ref) {
    debug('refd socket');
    this._socket.ref();
  } else if (this._timer) {
    clearTimeout(this._timer);
    this._timer = null;
  }
};

SocketQueue.prototype._onlisten = function() {
  this._unref();
  this._listening = true;
  this._fill();
};

SocketQueue.prototype._onclose = function() {
  var req, err, self = this;

  debug('socket closed', this);

  this._listening = false;

  err = new Error('getHostByName ' + consts.TIMEOUT);
  err.errno = consts.TIMEOUT;

  while (this._pending.length) {
    req = this._pending.pop();
    req.error(err);
  }

  Object.keys(this._active).forEach(function(key) {
    var req = self._active[key];
    req.error(err);
    delete self._active[key];
    self._active_count -= 1;
  });
};

var serverHash = function(server) {
  if (server.type === 'tcp')
    return server.address + ':' + server.port;
  else
    return 'udp' + net.isIP(server.address);
};

var _sockets = {};

exports.send = function(request) {
  var hash = serverHash(request.server);
  var socket = _sockets[hash];

  if (!socket) {
    switch (hash) {
      case 'udp4':
      case 'udp6':
        socket = new SocketQueue(new UDPSocket(), hash);
        break;
      default:
        socket = new SocketQueue(new TCPSocket(), request.server);
        break;
    }

    socket.on('close', function() {
      delete _sockets[hash];
    });

    _sockets[hash] = socket;
  }

  socket.send(request);
};

exports.remove = function(request) {
  var hash = serverHash(request.server);
  var socket = _sockets[hash];
  if (socket) {
    socket.remove(request);
  }
};


lib/utils.js
// Copyright 2012 Timothy J Fontaine <tjfontaine@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE

var dgram = require('dgram');
var EventEmitter = require('events').EventEmitter;
var ipaddr = require('ipaddr.js');
var net = require('net');
var util = require('util');
var fs = require('fs');

var UDPSocket = exports.UDPSocket = function(socket, remote) {
  this._socket = socket;
  this._remote = remote;
  this._buff = undefined;
  this.base_size = 512;
  this.bound = false;
  this.unref = undefined;
  this.ref = undefined;
};
util.inherits(UDPSocket, EventEmitter);

UDPSocket.prototype.buffer = function(size) {
  this._buff = new Buffer(size);
  return this._buff;
};

UDPSocket.prototype.send = function(len) {
  this._socket.send(this._buff, 0, len, this._remote.port,
                    this._remote.address);
};

UDPSocket.prototype.bind = function(type) {
  var self = this;

  if (this.bound) {
    this.emit('ready');
  } else {
    this._socket = dgram.createSocket(type);
    this._socket.on('listening', function() {
      self.bound = true;
      if (self._socket.unref) {
        self.unref = function() {
          self._socket.unref();
        };
        self.ref = function() {
          self._socket.ref();
        };
      }
      self.emit('ready');
    });

    this._socket.on('message', this.emit.bind(this, 'message'));

    this._socket.on('close', function() {
      self.bound = false;
      self.emit('close');
    });

    this._socket.bind();
  }
};

UDPSocket.prototype.close = function() {
  this._socket.close();
};

UDPSocket.prototype.remote = function(remote) {
  return new UDPSocket(this._socket, remote);
};

var TCPSocket = exports.TCPSocket = function(socket) {
  UDPSocket.call(this, socket);
  this.base_size = 4096;
  this._rest = undefined;
};
util.inherits(TCPSocket, UDPSocket);

TCPSocket.prototype.buffer = function(size) {
  this._buff = new Buffer(size + 2);
  return this._buff.slice(2);
};

TCPSocket.prototype.send = function(len) {
  this._buff.writeUInt16BE(len, 0);
  this._socket.write(this._buff.slice(0, len + 2));
};

TCPSocket.prototype.bind = function(server) {
  var self = this;

  if (this.bound) {
    this.emit('ready');
  } else {
    this._socket = net.connect(server.port, server.address);

    this._socket.on('connect', function() {
      self.bound = true;
      if (self._socket.unref) {
        self.unref = function() {
          self._socket.unref();
        };
        self.ref = function() {
          self._socket.ref();
        };
      }
      self.emit('ready');
    });

    this._socket.on('timeout', function() {
      self.bound = false;
      self.emit('close');
    });

    this._socket.on('close', function() {
      self.bound = false;
      self.emit('close');
    });

    this.catchMessages();
  }
};

TCPSocket.prototype.catchMessages = function() {
  var self = this;
  this._socket.on('data', function(data) {
    var len, tmp;
    if (!self._rest) {
      self._rest = data;
    } else {
      tmp = new Buffer(self._rest.length + data.length);
      self._rest.copy(tmp, 0);
      data.copy(tmp, self._rest.length);
      self._rest = tmp;
    }
    while (self._rest && self._rest.length > 2) {
      len = self._rest.readUInt16BE(0);
      if (self._rest.length >= len + 2) {
        self.emit('message', self._rest.slice(2, len + 2), self);
        self._rest = self._rest.slice(len + 2);
      } else {
        break;
      }
    }
  });
};

TCPSocket.prototype.close = function() {
  this._socket.end();
};

TCPSocket.prototype.remote = function() {
  return this;
};

exports.reverseIP = function(ip) {
  var address, kind, reverseip, parts;
  address = ipaddr.parse(ip.split(/%/)[0]);
  kind = address.kind();

  switch (kind) {
    case 'ipv4':
      address = address.toByteArray();
      address.reverse();
      reverseip = address.join('.') + '.IN-ADDR.ARPA';
      break;
    case 'ipv6':
      parts = [];
      address.toNormalizedString().split(':').forEach(function(part) {
        var i, pad = 4 - part.length;
        for (i = 0; i < pad; i++) {
          part = '0' + part;
        }
        part.split('').forEach(function(p) {
          parts.push(p);
        });
      });
      parts.reverse();
      reverseip = parts.join('.') + '.IP6.ARPA';
      break;
  }

  return reverseip;
};

exports.parseResolv = function parseResolv() {
  var nameservers = [];
  var file = fs.readFileSync('/etc/resolv.conf', 'ascii');

  file.split(/\n/).forEach(function(line) {
    var parts;
    line = line.replace(/^\s+|\s+$/g, '');
    if (!line.match(/^#/)) {
      parts = line.split(/\s+/);
      // only pick nameserver field
      if (parts[0] === 'nameserver') {
        nameservers.push(parts[1]);
      }
    }
  });

  return nameservers;
};
